File: ml-benchmark/synduce/constraints/sortedlist/mss.f
Inductive List = elt Int | cons {Int, List};
Inductive CList = single Int | concat {Int, CList, CList};

cat = fix (
  \f: List -> List -> List. \a: List. \b: List.
  match a with
    elt w -> cons {w, b}
  | cons {h, t} -> cons {h, (f t b)}
  end
);

repr = fix (
  \f: CList -> List. \c: CList.
  match c with
    single w -> elt w
  | concat {w, l, r} -> cat (f l) (f r)
  end
);

max = \a: Int. \b: Int. if < a b then b else a;
min = \a: Int. \b: Int. if < a b then a else b;

lmax = fix (
  \f: CList -> Int. \c: CList.
  match c with
    single w -> w
  | concat {w, l, r} -> max (f l) (f r)
  end
);

lmin = fix (
  \f: CList -> Int. \c: CList.
  match c with
    single w -> w
  | concat {w, l, r} -> min (f l) (f r)
  end
);

is_parti = fix (
  \f: CList -> Bool. \c: CList.
  match c with
    single w -> true
  | concat {w, l, r} -> and (and (< (lmax l) w) (< w (lmin r))) (and (f l) (f r))
  end
);

spec = fix (
  \f: List -> {Int, Int, Int, Int}. \xs: List.
  match xs with
    elt w -> {w, max w 0, max w 0, max w 0}
  | cons {h, t} ->
    let res = f t in
      {+ res.1 h, max res.2 (+ res.1 h), max (+ res.3 h) 0, max res.4 (+ res.3 h)}
  end
);

sum = fix (
  \f: CList -> Int. \c: CList.
  match c with
    single a -> a
  | concat {a, l, r} -> + (f l) (f r)
  end
);

target = fix (
  \f: CList -> Compress CList. \xs: CList.
  match xs with
    single a -> xs
  | concat {w, l, r} ->
    if < w 0 then
      let s = sum l in /*This invocation is provided in Synduce's template*/
        concat {w, l, f r} /*Avoid the recursion of l*/
    else concat {w, f l, f r}
  end
);

/*Customized Generator*/

insert = \w: Int. fix (
  \f: List -> List. \xs: List.
  match xs with
    elt a -> if < w a then cons {w, elt a} else cons {a, elt w}
  | cons {h, t} ->
    if < w h then cons {w, xs} else cons {h, f t}
  end
);

sort = fix (
  \f: List -> List. \xs: List.
  match xs with
    elt w -> xs
  | cons {h, t} -> insert h (f t)
  end
);

access = \x: List. match x with
  elt w -> {w, x}
| cons {h, t} -> {h, t}
end;

fill = fix (
  \f: CList -> List -> {CList, List}. \c: CList. \xs: List.
  match c with
    single _ ->
      let info = access xs in
        {single info.1, info.2}
  | concat {_, l, r} ->
    let lres = f l xs in
      let info = access lres.2 in
        let rres = f r info.2 in
          {concat {info.1, lres.1, rres.1}, rres.2}
  end
);

flatten = fix (
  \f: CList -> List. \c: CList.
  match c with
    single w -> elt w
  | concat {w, l, r} -> cat (f l) (cons {w, (f r)})
  end
);

gen = \c: CList. let xs = sort (flatten c) in
  (fill c xs).1;

main = \xs: CList.
  let inp = gen xs in
    if is_parti inp then spec (repr (target inp)) else {0, 0, 0, 0};

File: ml-benchmark/synduce/constraints/sortedlist/mss.ml
type list_ = Elt of int | Cons of int * list_
type clist = Single of int | Concat of int * clist * clist

let rec cat a b =
  match a with
  | Elt w -> Cons (w, b)
  | Cons (h, t) -> Cons (h, cat t b)

let rec repr c =
  match c with
  | Single w -> Elt w
  | Concat (_, l, r) -> cat (repr l) (repr r)

let max a b =
  if a < b then b else a

let min a b =
  if a < b then a else b

let rec lmax c =
  match c with
  | Single w -> w
  | Concat (_, l, r) -> max (lmax l) (lmax r)

let rec lmin c =
  match c with
  | Single w -> w
  | Concat (_, l, r) -> min (lmin l) (lmin r)

let rec is_parti c =
  match c with
  | Single _ -> true
  | Concat (w, l, r) -> (lmax l) < w && w < (lmin r) && (is_parti l && is_parti r)

let rec spec xs =
  match xs with
  | Elt w -> (w, max w 0, max w 0, max w 0)
  | Cons (h, t) ->
      let res = spec t in
      match res with
      | (a1, a2, a3, a4) ->
          let s1 = a1 + h in
          let s2 = max a2 s1 in
          let s3 = max (a3 + h) 0 in
          let s4 = max a4 (a3 + h) in
          (s1, s2, s3, s4)

let rec sum c =
  match c with
  | Single a -> a
  | Concat (_, l, r) -> sum l + sum r

val target: clist -> clist compress
let rec target xs =
  match xs with
  | Single _ -> xs
  | Concat (w, l, r) ->
      if w < 0 then
        let s = sum l in
        Concat (w, l, target r)
      else
        Concat (w, target l, target r)

let rec insert w xs =
  match xs with
  | Elt a ->
      if w < a then Cons (w, Elt a) else Cons (a, Elt w)
  | Cons (h, t) ->
      if w < h then Cons (w, xs) else Cons (h, insert w t)

let rec sort xs =
  match xs with
  | Elt _ -> xs
  | Cons (h, t) -> insert h (sort t)

let access x =
  match x with
  | Elt w -> (w, x)
  | Cons (h, t) -> (h, t)

let rec fill c xs =
  match c with
  | Single _ ->
      let info = access xs in (
      match info with
      | (i1, i2) -> (Single i1, i2)
      )
  | Concat (_, l, r) ->
      let lres = fill l xs in
      match lres with
      | (lcl, lxs_tail) ->
          let info = access lxs_tail in (
          match info with
          | (i1, i2) ->
            let rres = fill r i2 in (
              match rres with
              | (rcl, rxs_tail) -> (Concat (i1, lcl, rcl), rxs_tail)
            )
          )

let rec flatten c =
  match c with
  | Single w -> Elt w
  | Concat (w, l, r) -> cat (flatten l) (Cons (w, flatten r))

let gen c =
  let xs = sort (flatten c) in
  match fill c xs with
  | (cl, _) -> cl

let program xs =
  let inp = gen xs in
  if is_parti inp then spec (repr (target inp)) else (0, 0, 0, 0)